src/StoreApi/Routes/V1/AbstractCartRoute.php

```diff
@@ -99,13 +102,65 @@ public function get_response( \WP_REST_Request $request ) {
 	 * @param \WP_REST_Response $response The response object.
 	 * @return \WP_REST_Response
 	 */
-	protected function add_nonce_headers( \WP_REST_Response $response ) {
-		$response->header( 'X-WC-Store-API-Nonce', wp_create_nonce( 'wc_store_api' ) );
-		$response->header( 'X-WC-Store-API-Nonce-Timestamp', time() );
-		$response->header( 'X-WC-Store-API-User', get_current_user_id() );
+	protected function add_response_headers( \WP_REST_Response $response ) {
+		$response->header( 'Nonce', wp_create_nonce( 'wc_store_api' ) );
+		$response->header( 'Nonce-Timestamp', time() );
+		$response->header( 'User-ID', get_current_user_id() );
+		$response->header( 'Cart-Token', $this->get_cart_token( wc()->session->get_customer_id() ) );
 		return $response;
 	}
 
+	/**
+	 * Load the cart session before handling responses.
+	 *
+	 * @param \WP_REST_Request $request Request object.
+	 */
+	protected function load_cart_session( \WP_REST_Request $request ) {
+		$cart_token = $request->get_header( 'Cart-Token' );
+
+		if ( $cart_token && Token::validate( $cart_token, '@' . wp_salt() ) ) {
+			// Overrides the core session class.
+			add_filter(
+				'woocommerce_session_handler',
+				function() {
+					return SessionHandler::class;
+				}
+			);
+		}
+
+		$this->cart_controller->load_cart();
+	}
+
+	/**
+	 * Generates a cart token for the response headers.
+	 *
+	 * Current namespace is used as the token Issuer.
+	 *
+	 * @param string $customer_id Session (customer ID) for the token.
+	 * @return string
+	 */
+	protected function get_cart_token( $customer_id ) {
+		return Token::create( $customer_id, $this->get_cart_token_secret(), $this->get_cart_token_expiration(), $this->namespace );
```

Comment:
Is it strictly necessary to use a JWT library to create this useful `Cart-Token` header?

Could the existing approach from [set_customer_session_cookie](https://github.com/woocommerce/woocommerce/blob/9db2c452905ebf796a8e8b4d63db64b57e169dbe/plugins/woocommerce/includes/class-wc-session-handler.php#L160-L161) be used?

Given the session cookie is already present in the API response headers, providing the same data in the `Cart-Token` would be consistent.

Reply:
Hi Anthony, I picked up this PR. I agree we should get rid of this library if we can. Also, we're using it with `wp_salt()` as the secret and it enforces a "secretValidation" (🙄) causing it to throw an error if our provided secret does not obey certain security rules (either default or provided by us). And nowhere it is certain that `wp_salt()` will, I don't believe we enforce salt keys rules right?

I'm having trouble locating the header you mentioned in API responses, can you send me an example of the response header you are talking about, or point out where is it inserted?

Reply:
I left [some feedback](https://github.com/woocommerce/woocommerce-blocks/issues/5683#issuecomment-1158771475) on the Issue thread about this.